File: greg_and_graph.cpp
#include <iostream>
#include <algorithm>
typedef long long ll;
using namespace std;

int main(){
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  int n; cin>>n;
  ll adjmat[n][n];
  int order[n];
  ll ans[n];
  bool allowed[n];
  for(int i=0;i<n;i++)
    for(int j=0;j<n;j++)
      cin>>adjmat[i][j];

  for(int i=0;i<n;i++){
    cin>>order[n-i-1];
    ans[i] = 0;
    allowed[i] = false;
  }

  for(int k=0;k<n;k++){
    int added = order[k]-1;
    allowed[added] = true;
    for(int i=0;i<n;i++){
      for(int j=0;j<n;j++){
        adjmat[i][j] = min(adjmat[i][j], adjmat[i][added] + adjmat[added][j]);
        if (i!=j && allowed[i] && allowed[j])
          ans[n-k-1]+=adjmat[i][j];
      }
    }
  }
  for(ll u: ans) cout<<u<<" ";
}

File: greg_and_graph.py
n = int(input())
adjmat = [list(map(int, input().split())) for i in range(n)]
order = list(map(int, input().split()))
order.reverse()
ans = []
allowed = [False]*n
for k in order:
    allowed[k-1] = True
    part = 0
    for i in range(n):
        for j in range(n):
            x = i+1
            y = j+1
            adjmat[i][j] = min(adjmat[i][j], adjmat[i][k-1]+adjmat[k-1][j])
            if x!=y and (allowed[i] and allowed[j]):
                part += adjmat[i][j]
    ans.append(part)
ans.reverse()
print(' '.join(list(map(str, ans))))
